### PR TITLE
[ENG-4157] Add mirage for osf api

### PR DESCRIPTION
**Pull request by brianjgeiger**
-   Ticket: [ENG-4157]

## Purpose

Provide mirage endpoints for the OSF parts of the metadata project. The way this works is that, any time a guid is created, the appropriate custom-metadata record will also be created to go with. This means that if you navigate to, for example, a registration via the my-registrations page, a metadata record will not be created, because it won't go via the guid route, and a guid won't be created for it.

We _could_ automatically create guids for all records, but I think that would be a bad idea overall. It would make two related records for anything that we have that's a guid-based object, and that would likely slow down mirage a bit. It's probably best to just know that you may have to reload the page to get your metadata record when you're testing metadata things.

## Summary of Changes

1. Add factories for the custom metadata records
2. Add a factory for Guids to automatically add metadata records to the guid
3. Make the guid creation use mirage so that it activates the factory
4. Add endpoints to the mriage config
5. Serialize the `custom_metadata` relationship in the guid
6. Add serializers for custom metadata records to fix `self` link and add `guid` relationship

## Screenshot(s)
Screen Shot 2022-11-04 at 11 41 12 AM

Screen Shot 2022-11-04 at 11 40 29 AM

Screen Shot 2022-11-04 at 11 42 06 AM

## Side Effects

No

## QA Notes

Still not ready for QA

[ENG-4157]: https://openscience.atlassian.net/browse/ENG-4157?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: mirage/config.ts [hunk 1]
```diff
@@ -318,6 +318,15 @@ export default function(this: Server) {
     });
     this.post('/search/collections/', searchCollections);
 
+    osfResource(this, 'custom-file-metadata-record', {
+        only: ['show', 'update'],
+        path: '/custom-file-metadata-records',
```
Comment (adlius) [inline mirage/config.ts:323]: In the urls, our API usually use snake case. Are we sure the path is not `/custom_file_metadata_records`?

File: mirage/config.ts [hunk 2]
```diff
@@ -318,6 +318,15 @@ export default function(this: Server) {
...
+    });
+    osfResource(this, 'custom-item-metadata-record', {
+        only: ['show', 'update'],
+        path: '/custom-item-metadata-records',
```
Comment (adlius) [inline mirage/config.ts:327]: Same here.

File: mirage/factories/custom-file-metadata-record.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+export default Factory.extend<CustomFileMetadataRecord>({
+    title: faker.lorem.sentence(),
+    description: faker.lorem.sentence(),
+    resource_type_general: 'text',
+    language: 'en-US',
+});
+
+declare module 'ember-cli-mirage/types/registries/model' {
+    export default interface MirageModelRegistry {
+        customFileMetadataRecords: CustomFileMetadataRecord;
```
Comment (adlius) [inline mirage/factories/custom-file-metadata-record.ts:16]: If you look at our other mirage factories (`schema-response`, `review-action`), it seems that for the `MirageModelRegistry` we should do:
```suggestion
        'custom-file-metadata-record': CustomFileMetadataRecord;
```

File: mirage/factories/custom-item-metadata-record.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            award_numer: '67890',
+            award_uri: faker.internet.url,
+            award_title: faker.lorem.sentence,
+        },
+    ],
+});
+
+declare module 'ember-cli-mirage/types/registries/model' {
+    export default interface MirageModelRegistry {
+        customItemMetadataRecords: CustomItemMetadataRecord;
```
Comment (adlius) [inline mirage/factories/custom-item-metadata-record.ts:32]: Same as above.

File: mirage/serializers/guid.ts [hunk 5]
```diff
@@ -16,6 +16,8 @@ export default class GuidSerializer extends ApplicationSerializer<SerializedGuid
         const pluralizedType = pluralize(guid.referentType!);
         const referent = guid._schema[pluralizedType].find(guid.id);
         const typeKey = this.typeKeyForModel(referent);
+        const metadataType = guid.referentType === 'file' ?
+            'custom-file-metadata-records' : 'custom-item-metadata-records';
```
Comment (adlius) [inline mirage/serializers/guid.ts:20]: Should this be snake case?

**Review by adlius — CHANGES_REQUESTED** [review]
Some questions about URL snake_case vs kebab-case and two small changes.

**Review by chth0n1x — APPROVED** [review]
These look good, approving.

File: mirage/factories/custom-item-metadata-record.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+            award_numer: '67890',
+            award_uri: faker.internet.url,
+            award_title: faker.lorem.sentence,
+        },
+    ],
+});
+
+declare module 'ember-cli-mirage/types/registries/model' {
+    export default interface MirageModelRegistry {
+        'custom-item-metadata-records': CustomItemMetadataRecord;
```
Comment (adlius) [inline mirage/factories/custom-item-metadata-record.ts:32]:
```suggestion
        'custom-item-metadata-record': CustomItemMetadataRecord;
```

File: mirage/factories/guid.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    afterCreate(guid, server) {
+        const metadataType = guid.referentType === 'file' ?
+            'custom-file-metadata-record' : 'custom-item-metadata-record';
+        server.create(metadataType, { id: guid.id });
+    },
+});
+
+declare module 'ember-cli-mirage/types/registries/model' {
+    export default interface MirageModelRegistry {
+        guids: GuidModel;
```
Comment (adlius) [inline mirage/factories/guid.ts:15]:
```suggestion
        guid: GuidModel;
```

File: mirage/factories/custom-file-metadata-record.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+export default Factory.extend<CustomFileMetadataRecord>({
+    title: faker.lorem.sentence(),
+    description: faker.lorem.sentence(),
+    resource_type_general: 'text',
+    language: 'en-US',
+});
+
+declare module 'ember-cli-mirage/types/registries/model' {
+    export default interface MirageModelRegistry {
+        'custom-file-metadata-records': CustomFileMetadataRecord;
```
Comment (adlius) [inline mirage/factories/custom-file-metadata-record.ts:16]:
```suggestion
        'custom-file-metadata-record': CustomFileMetadataRecord;
```

**Review by adlius — CHANGES_REQUESTED** [review]

**Review by adlius — APPROVED** [review]

**Pull request merged**